### PR TITLE
[SPARK-37259][SQL] Support CTE and temp table queries with MSSQL JDBC

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
Currently CTE queries from Spark are not supported with MSSQL server via JDBC. This is because MSSQL server doesn't support the nested CTE syntax (`SELECT * FROM (WITH t AS (...) SELECT ... FROM t) WHERE 1=0`) that Spark builds from the original query (`options.tableOrQuery`) in `JDBCRDD.resolveTable()`  and in `JDBCRDD.compute()`.
Unfortunately, it is non-trivial to split an arbitrary query into "with" and "regular" clauses in `MsSqlServerDialect`. So instead, I'm proposing a new general JDBC option `prepareQuery` that users can use if they have complex queries:
```
val df = spark.read.format("jdbc")
  .option("url", jdbcUrl)
  .option("prepareQuery", "WITH t AS (SELECT x, y FROM tbl)")
  .option("query", "SELECT * FROM t WHERE x > 10")
  .load()
```
This change also works with MSSQL's temp table syntax:
```
val df = spark.read.format("jdbc")
  .option("url", jdbcUrl)
  .option("prepareQuery", "(SELECT * INTO #TempTable FROM (SELECT * FROM tbl WHERE x > 10) t)")
  .option("query", "SELECT * FROM #TempTable")
  .load()
```

### Why are the changes needed?
To support CTE and temp table queries with MSSQL.

### Does this PR introduce _any_ user-facing change?
Yes, CTE and temp table queries are supported form now.

### How was this patch tested?
Added new integration UTs.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @cloud-fan, @KevinAppelBofa, @WaterKnight1998, @huaxingao, @beliefer

**Comment by KevinAppelBofa** [comment]
Thank you @peter-toth, I did a git clone and grabbed the latest source and compiled, this is showing 3.4.0-SNAPSHOT.  I added in your changes and recompiled the spark-sql and then ran some of the tests.  I also ran our actual real sql server queries using this and everything is working correctly.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala [hunk 1]
```diff
@@ -374,4 +375,58 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationSuite {
     val filtered = df.where(col("c") === 0).collect()
     assert(filtered.length == 0)
   }
+
+  test("prepareQuery and query JDBC options") {
```
Comment (huaxingao) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:379]: Could you put the jira number here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 2]
```diff
@@ -222,6 +222,8 @@ class JDBCOptions(
 
   // User specified JDBC connection provider name
   val connectionProviderName = parameters.get(JDBC_CONNECTION_PROVIDER)
+
+  val prepareQuery = parameters.get(JDBC_PREPARE_QUERY).map(_ + " ").getOrElse("")
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:226]: add a brief comment to explain what this is for?

**Comment by huaxingao** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:226]: +1

**Review by cloud-fan — APPROVED** [review]

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala (hunk 1 above)
Comment (peter-toth) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala]: Added in https://github.com/apache/spark/pull/36440/commits/aca8fa236f2de5144eb68e970f3b00255433ab7b

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 2 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:228]: Fixed in https://github.com/apache/spark/pull/36440/commits/aca8fa236f2de5144eb68e970f3b00255433ab7b

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by huaxingao** [comment]
LGTM. Merged to master.
Thanks @peter-toth for fixing this and thanks @cloud-fan @dongjoon-hyun @KevinAppelBofa for reviewing and verifying.

**Comment by peter-toth** [comment]
Thanks all for the review and comments.

**Pull request closed without merging**